app/src/shared/components/TextControl/index.jsx

```diff
@@ -0,0 +1,158 @@
+// @flow
+
+import React, { useCallback, type Node, useRef, useState, useEffect } from 'react'
+import { type Ref } from '$shared/flowtype/common-types'
+
+type TextControlRef = Ref<HTMLTextAreaElement | HTMLInputElement>
+
+type Props = {
+    commitEmpty?: boolean,
+    flushHistoryOnBlur?: boolean,
+    immediateCommit?: boolean,
+    innerRef?: ?TextControlRef,
+    onBlur?: ?(SyntheticInputEvent<EventTarget>) => void,
+    onChange?: ?(SyntheticInputEvent<EventTarget>) => void,
+    onCommit?: ?(string) => void,
+    onFocus?: ?(SyntheticInputEvent<EventTarget>) => void,
+    onKeyDown?: ?(SyntheticKeyboardEvent<EventTarget>) => void,
+    revertOnEsc?: boolean,
+    selectAllOnFocus?: boolean,
+    tag?: 'input' | 'textarea',
+    value?: string,
+}
+
+const normalize = (value: any): string => (
+    typeof value === 'string' ? value.trim() : String(value)
+)
+
+const TextControl = ({
+    commitEmpty,
+    flushHistoryOnBlur,
+    immediateCommit,
+    innerRef,
+    onBlur: onBlurProp,
+    onChange: onChangeProp,
+    onCommit,
+    onFocus: onFocusProp,
+    onKeyDown: onKeyDownProp,
+    revertOnEsc,
+    selectAllOnFocus,
+    tag,
+    value: valueProp,
+    ...props
+}: Props): Node => {
+    const Tag = tag || TextControl.defaultProps.tag
+    const el = useRef(null)
+    const ref = innerRef || el
+    const reverted: Ref<boolean> = useRef(false)
+    const [value, setValue] = useState(valueProp || '')
+    const [blurCount, setBlurCount] = useState(0)
+    const normalizedValue = normalize(value)
+    const commit = useCallback(() => {
+        if (onCommit && normalizedValue !== normalize(valueProp) && (normalizedValue || commitEmpty)) {
+            onCommit(normalizedValue)
+        }
+    }, [normalizedValue, valueProp, onCommit, commitEmpty])
+
+    const onBlur = useCallback((e: SyntheticInputEvent<EventTarget>) => {
+        if (flushHistoryOnBlur) {
+            setBlurCount(blurCount + 1)
```

Comment:
Probably warrants a comment as it's not at all obvious what this is for/how/why this works.